### PR TITLE
Issue #18: Ssh Protocol Check Metric

### PR DESCRIPTION
**Pull request by CherfaElyes**
* Developped SSH protocol health check
* Created a test case for SSH protocol health check
* Tested SSH protocol health check on metricshub with the agent

### CODE REVIEW

**Review by NassimBtk — CHANGES_REQUESTED** [review]
Missing OsCommand remote and local executions  management.

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java [hunk 1]
```diff
@@ -97,6 +104,52 @@ private TelemetryManager createTelemetryManagerWithSnmpConfig() {
...
+		// Create a telemetry manager
+		return TelemetryManager
+			.builder()
+			.monitors(monitors)
+			.hostConfiguration(
+				HostConfiguration
+					.builder()
+					.hostId(HOSTNAME)
+					.hostname(HOSTNAME)
+					.sequential(false)
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java:122]: sequential is not required in all the telemetry managers.

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java [hunk 2]
```diff
@@ -184,4 +237,216 @@ void testCheckSnmpDownHealth() throws InterruptedException, ExecutionException,
 
 		assertEquals(DOWN, telemetryManager.getEndpointHostMonitor().getMetric(SNMP_UP_METRIC).getValue());
 	}
+
+	@Test
+	void testCheckSshHealthLocally() {
+		// Create a telemetry manager using an SNMP HostConfiguration.
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java:243]: `// Create a telemetry manager using an SSH HostConfiguration.`

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java [hunk 3]
```diff
@@ -184,4 +237,216 @@ void testCheckSnmpDownHealth() throws InterruptedException, ExecutionException,
...
+		// Setting SSH host properties
+		telemetryManager.getHostProperties().setMustCheckSshStatus(true);
+		telemetryManager.getHostProperties().setOsCommandExecutesLocally(true);
+
+		try (MockedStatic<OsCommandHelper> staticOsCommandHelper = Mockito.mockStatic(OsCommandHelper.class)) {
+			staticOsCommandHelper
+				.when(() -> OsCommandHelper.runLocalCommand(anyString(), anyLong(), any()))
+				.thenReturn(SUCCESS_RESPONSE);
+
+			// Start the SNMP Health Check strategy
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java:262]: `// Start the SSH Health Check strategy`

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java [hunk 4]
```diff
@@ -184,4 +237,216 @@ void testCheckSnmpDownHealth() throws InterruptedException, ExecutionException,
...
+			sshHealthCheckStrategy.run();
+
+			assertEquals(DOWN, telemetryManager.getEndpointHostMonitor().getMetric(SSH_UP_METRIC).getValue());
+		}
+	}
+
+	@Test
+	void testCheckSshUpHealthRemotely() {
+		// Create a telemetry manager using an SNMP HostConfiguration.
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java:280]: `// Create a telemetry manager using an SSH HostConfiguration.`

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java [hunk 5]
```diff
@@ -184,4 +237,216 @@ void testCheckSnmpDownHealth() throws InterruptedException, ExecutionException,
...
+		telemetryManager.getHostProperties().setOsCommandExecutesRemotely(true);
+
+		try (MockedStatic<OsCommandHelper> staticOsCommandHelper = Mockito.mockStatic(OsCommandHelper.class)) {
+			staticOsCommandHelper
+				.when(() ->
+					OsCommandHelper.runSshCommand(anyString(), anyString(), any(SshConfiguration.class), anyLong(), any(), any())
+				)
+				.thenReturn(SUCCESS_RESPONSE);
+
+			// Start the SNMP Health Check strategy
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java:302]: `// Start the SSH Health Check strategy`

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java [hunk 6]
```diff
@@ -184,4 +237,216 @@ void testCheckSnmpDownHealth() throws InterruptedException, ExecutionException,
...
+		}
+
+		try (MockedStatic<OsCommandHelper> staticOsCommandHelper = Mockito.mockStatic(OsCommandHelper.class)) {
+			staticOsCommandHelper
+				.when(() ->
+					OsCommandHelper.runSshCommand(anyString(), anyString(), any(SshConfiguration.class), anyLong(), any(), any())
+				)
+				.thenReturn(null);
+
+			// Start the SNMP Health Check strategy
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java:315]: `// Start the SSH Health Check strategy`

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java [hunk 7]
```diff
@@ -184,4 +237,216 @@ void testCheckSnmpDownHealth() throws InterruptedException, ExecutionException,
...
+			sshHealthCheckStrategy.run();
+
+			assertEquals(DOWN, telemetryManager.getEndpointHostMonitor().getMetric(SSH_UP_METRIC).getValue());
+		}
+	}
+
+	@Test
+	void testCheckSshUpHealthBothLocallyAndRemotely() {
+		// Create a telemetry manager using an SNMP HostConfiguration.
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java:324]: `// Create a telemetry manager using an SSH HostConfiguration.`

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java [hunk 8]
```diff
@@ -184,4 +237,216 @@ void testCheckSnmpDownHealth() throws InterruptedException, ExecutionException,
...
+				.when(() ->
+					OsCommandHelper.runSshCommand(anyString(), anyString(), any(SshConfiguration.class), anyLong(), any(), any())
+				)
+				.thenReturn(SUCCESS_RESPONSE);
+
+			staticOsCommandHelper
+				.when(() -> OsCommandHelper.runLocalCommand(anyString(), anyLong(), any()))
+				.thenReturn(SUCCESS_RESPONSE);
+
+			// Start the SNMP Health Check strategy
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java:351]: `// Start the SSH Health Check strategy`

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java [hunk 9]
```diff
@@ -184,4 +237,216 @@ void testCheckSnmpDownHealth() throws InterruptedException, ExecutionException,
...
+		try (MockedStatic<OsCommandHelper> staticOsCommandHelper = Mockito.mockStatic(OsCommandHelper.class)) {
+			staticOsCommandHelper
+				.when(() ->
+					OsCommandHelper.runSshCommand(anyString(), anyString(), any(SshConfiguration.class), anyLong(), any(), any())
+				)
+				.thenReturn(SUCCESS_RESPONSE);
+
+			staticOsCommandHelper.when(() -> OsCommandHelper.runLocalCommand(anyString(), anyLong(), any())).thenReturn(null);
+
+			// Start the SNMP Health Check strategy
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java:367]: `// Start the SSH Health Check strategy`

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java [hunk 10]
```diff
@@ -184,4 +237,216 @@ void testCheckSnmpDownHealth() throws InterruptedException, ExecutionException,
...
+		try (MockedStatic<OsCommandHelper> staticOsCommandHelper = Mockito.mockStatic(OsCommandHelper.class)) {
+			staticOsCommandHelper
+				.when(() ->
+					OsCommandHelper.runSshCommand(anyString(), anyString(), any(SshConfiguration.class), anyLong(), any(), any())
+				)
+				.thenReturn(null);
+
+			staticOsCommandHelper.when(() -> OsCommandHelper.runLocalCommand(anyString(), anyLong(), any())).thenReturn(null);
+
+			// Start the SNMP Health Check strategy
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java:400]: `// Start the SSH Health Check strategy`

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java [hunk 11]
```diff
@@ -184,4 +237,216 @@ void testCheckSnmpDownHealth() throws InterruptedException, ExecutionException,
...
+			sshHealthCheckStrategy.run();
+
+			assertEquals(DOWN, telemetryManager.getEndpointHostMonitor().getMetric(SSH_UP_METRIC).getValue());
+		}
+	}
+
+	@Test
+	void testCheckSshNoHealthWhenMustCheckFalse() {
+		// Create a telemetry manager using an SNMP HostConfiguration.
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java:409]: `// Create a telemetry manager using an SSH HostConfiguration.`

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java [hunk 12]
```diff
@@ -184,4 +237,216 @@ void testCheckSnmpDownHealth() throws InterruptedException, ExecutionException,
...
+			CURRENT_TIME_MILLIS,
+			clientsExecutorMock
+		);
+
+		// Setting SSH host properties
+		telemetryManager.getHostProperties().setMustCheckSshStatus(false);
+		telemetryManager.getHostProperties().setOsCommandExecutesLocally(true);
+		telemetryManager.getHostProperties().setOsCommandExecutesRemotely(true);
+
+		// Start the SNMP Health Check strategy
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java:424]: `// Start the SSH Health Check strategy`

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java [hunk 13]
```diff
@@ -184,4 +237,216 @@ void testCheckSnmpDownHealth() throws InterruptedException, ExecutionException,
...
+		sshHealthCheckStrategy.run();
+
+		assertNull(telemetryManager.getEndpointHostMonitor().getMetric(SSH_UP_METRIC));
+	}
+
+	@Test
+	void testCheckSshNoHealthWhenNoConfiguration() {
+		// Create a telemetry manager using an SNMP HostConfiguration.
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java:432]: `// Create a telemetry manager without configuration.`

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java [hunk 14]
```diff
@@ -184,4 +237,216 @@ void testCheckSnmpDownHealth() throws InterruptedException, ExecutionException,
...
+			CURRENT_TIME_MILLIS,
+			clientsExecutorMock
+		);
+
+		// Setting SSH host properties
+		telemetryManager.getHostProperties().setMustCheckSshStatus(false);
+		telemetryManager.getHostProperties().setOsCommandExecutesLocally(true);
+		telemetryManager.getHostProperties().setOsCommandExecutesRemotely(true);
+
+		// Start the SNMP Health Check strategy
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java:447]: `// Start the SSH Health Check strategy`

File: metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java [hunk 15]
```diff
@@ -184,4 +237,216 @@ void testCheckSnmpDownHealth() throws InterruptedException, ExecutionException,
...
+				.when(() ->
+					OsCommandHelper.runSshCommand(anyString(), anyString(), any(SshConfiguration.class), anyLong(), any(), any())
+				)
+				.thenReturn(null);
+
+			staticOsCommandHelper
+				.when(() -> OsCommandHelper.runLocalCommand(anyString(), anyLong(), any()))
+				.thenReturn(SUCCESS_RESPONSE);
+
+			// Start the SNMP Health Check strategy
```
Comment (NassimBtk) [inline metricshub-engine/src/test/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategyTest.java:384]: `// Start the SSH Health Check strategy`

File: metricshub-engine/src/main/java/org/sentrysoftware/metricshub/engine/strategy/detection/DetectionStrategy.java [hunk 16]
```diff
@@ -240,4 +250,70 @@ public void createMonitor(final ConnectorTestResult connectorTestResult) {
 
 		collectConnectorStatus(connectorTestResult, connector, connectorId, monitor);
 	}
+
+	/**
+	 * Verify the given set of sources instances to check if they are OsCommandSources
+	 *
+	 * @param criteria Connector detection criteria list
```
Comment (NassimBtk) [inline metricshub-engine/src/main/java/org/sentrysoftware/metricshub/engine/strategy/detection/DetectionStrategy.java:257]: Javadoc is wrong.
`@param sourceTypes Connector source types`

File: metricshub-engine/src/main/java/org/sentrysoftware/metricshub/engine/strategy/detection/DetectionStrategy.java [hunk 17]
```diff
@@ -240,4 +250,70 @@ public void createMonitor(final ConnectorTestResult connectorTestResult) {
...
+	 */
+	void verifySshSources(final Set<Class<? extends Source>> sources) {
```
Comment (NassimBtk) [inline metricshub-engine/src/main/java/org/sentrysoftware/metricshub/engine/strategy/detection/DetectionStrategy.java:259]: `void verifySshSources(final Set<Class<? extends Source>> sourceTypes) {`

**Review by NassimBtk — CHANGES_REQUESTED** [review]

File: metricshub-engine/src/main/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategy.java [hunk 18]
```diff
@@ -216,4 +233,101 @@ public void checkSnmpHealth(String hostname, Monitor hostMonitor, MetricFactory
...
+		}
+
+		return sshResult;
+	}
+
+	/**
+	 * Performs a remote SSH test to determine whether the SSH protocol is UP in the given host.
+	 *
+	 * @param hostname          The hostname on which we perform health check
+	 * @param sshResult         The results that will be used to create protocol health check metric
```
[the remark this answers is not in the document]
Comment (CherfaElyes) [reply metricshub-engine/src/main/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategy.java:308]: There is a conflict between this variable name "previousSshStatus" and the one you chose in the method signature which was "osCommandStatus". I'll be using "previousSshStatus".

File: metricshub-engine/src/main/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategy.java [hunk 19]
```diff
@@ -216,4 +233,101 @@ public void checkSnmpHealth(String hostname, Monitor hostMonitor, MetricFactory
...
+	 * @param sshConfiguration  The SSH configuration retrieved from the telemetryManager
+	 * @return The SSH health check result after performing the tests
+	 */
+	private Double remoteSshTest(String hostname, Double sshResult, SshConfiguration sshConfiguration) {
```
[the remark this answers is not in the document]
Comment (NassimBtk) [reply metricshub-engine/src/main/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategy.java]: private Double remoteSshTest(String hostname, Double previousSshStatus, SshConfiguration sshConfiguration) {

File: metricshub-engine/src/main/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategy.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (NassimBtk) [reply metricshub-engine/src/main/java/org/sentrysoftware/metricshub/engine/strategy/collect/ProtocolHealthCheckStrategy.java]: The right name is the one provided in the javadoc.

**Review by NassimBtk — CHANGES_REQUESTED** [review]
Additional changes are required.

**Review by NassimBtk — APPROVED** [review]

**Pull request merged**